rules/recording_test.go

```diff
@@ -83,6 +83,104 @@ func TestRuleEval(t *testing.T) {
+	exprWithoutMetricName, _ := parser.ParseExpr(`metric + metric`)
+
+	scenarios := []struct {
+		name       string
+		ruleLabels labels.Labels
+		expr       parser.Expr
+		expected   promql.Vector
+	}{
+		{
+			name:       "no labels in recording rule, metric name in query result",
+			ruleLabels: labels.EmptyLabels(),
+			expr:       exprWithMetricName,
+			expected: promql.Vector{
+				promql.Sample{
+					Metric: labels.FromStrings("__name__", "test_rule", "label_a", "1", "label_b", "3"),
+					Point:  promql.Point{V: 1, T: timestamp.FromTime(ts)},
+				},
+				promql.Sample{
+					Metric: labels.FromStrings("__name__", "test_rule", "label_a", "2", "label_b", "4"),
+					Point:  promql.Point{V: 1, T: timestamp.FromTime(ts)},
+				},
+			},
+		},
+		{
+			name:       "only new labels in recording rule, metric name in query result",
+			ruleLabels: labels.FromStrings("extra_from_rule", "foo"),
+			expr:       exprWithMetricName,
+			expected: promql.Vector{
+				promql.Sample{
+					Metric: labels.FromStrings("__name__", "test_rule", "label_a", "1", "label_b", "3", "extra_from_rule", "foo"),
+					Point:  promql.Point{V: 1, T: timestamp.FromTime(ts)},
+				},
+				promql.Sample{
+					Metric: labels.FromStrings("__name__", "test_rule", "label_a", "2", "label_b", "4", "extra_from_rule", "foo"),
+					Point:  promql.Point{V: 1, T: timestamp.FromTime(ts)},
+				},
+			},
+		},
+		{
+			name:       "some replacement labels in recording rule, metric name in query result",
+			ruleLabels: labels.FromStrings("label_a", "from_rule"),
+			expr:       exprWithMetricName,
+			expected: promql.Vector{
+				promql.Sample{
+					Metric: labels.FromStrings("__name__", "test_rule", "label_a", "from_rule", "label_b", "3"),
+					Point:  promql.Point{V: 1, T: timestamp.FromTime(ts)},
+				},
+				promql.Sample{
+					Metric: labels.FromStrings("__name__", "test_rule", "label_a", "from_rule", "label_b", "4"),
+					Point:  promql.Point{V: 1, T: timestamp.FromTime(ts)},
+				},
+			},
+		},
+		{
+			name:       "no labels in recording rule, no metric name in query result",
+			ruleLabels: labels.EmptyLabels(),
+			expr:       exprWithoutMetricName,
+			expected: promql.Vector{
+				promql.Sample{
+					Metric: labels.FromStrings("__name__", "test_rule", "label_a", "1", "label_b", "3"),
+					Point:  promql.Point{V: 2, T: timestamp.FromTime(ts)},
+				},
+				promql.Sample{
+					Metric: labels.FromStrings("__name__", "test_rule", "label_a", "2", "label_b", "4"),
+					Point:  promql.Point{V: 2, T: timestamp.FromTime(ts)},
+				},
+			},
+		},
+	}
+
+	for _, scenario := range scenarios {
+		b.Run(scenario.name, func(b *testing.B) {
+			rule := NewRecordingRule("test_rule", scenario.expr, scenario.ruleLabels)
+
+			b.ResetTimer()
+
+			for i := 0; i < b.N; i++ {
+				result, err := rule.Eval(suite.Context(), ts, EngineQueryFunc(suite.QueryEngine(), suite.Storage()), nil, 0)
+
+				require.NoError(b, err)
+				require.ElementsMatch(b, scenario.expected, result)
```

Comment:
Don't do a complex check in the inner-loop of a benchmark; it skews the result.
You could do it before `ResetTimer`, or perhaps better to add these cases to `TestRuleEval`.
I think the cases can be shared between Test and Benchmark.

Reply:
Good point, I've made this change in 054dfc3.

Reply:
The benchmark results now look like this, having re-benchmarked the original and new code with the modified benchmark:

```
name                                                                                old time/op    new time/op    delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                  7.07µs ± 2%    6.69µs ± 0%  -5.38%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10            7.21µs ± 0%    6.93µs ± 0%  -3.95%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10    7.05µs ± 0%    6.70µs ± 1%  -4.89%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10               14.3µs ± 1%    14.4µs ± 0%    ~     (p=0.690 n=5+5)

name                                                                                old alloc/op   new alloc/op   delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                  7.13kB ± 0%    6.44kB ± 0%  -9.66%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10            7.13kB ± 0%    6.83kB ± 0%  -4.26%  (p=0.029 n=4+4)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10    6.94kB ± 0%    6.44kB ± 0%  -7.16%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10               15.2kB ± 0%    14.8kB ± 0%  -2.83%  (p=0.008 n=5+5)

name                                                                                old allocs/op  new allocs/op  delta
RuleEval/no_labels_in_recording_rule,_metric_name_in_query_result-10                     127 ± 0%       119 ± 0%  -6.30%  (p=0.008 n=5+5)
RuleEval/only_new_labels_in_recording_rule,_metric_name_in_query_result-10               127 ± 0%       121 ± 0%  -4.72%  (p=0.008 n=5+5)
RuleEval/some_replacement_labels_in_recording_rule,_metric_name_in_query_result-10       125 ± 0%       119 ± 0%  -4.80%  (p=0.008 n=5+5)
RuleEval/no_labels_in_recording_rule,_no_metric_name_in_query_result-10                  239 ± 0%       233 ± 0%  -2.51%  (p=0.008 n=5+5)
```